File: 2019/levelUp_juneBatch/lecture_001/l002.cpp
#include <iostream>
#include <vector>
using namespace std;

int n, m;
vector<vector<int>> dirA{{0, 1}, {-1, 0}, {0, -1}, {1, 0}};
vector<string> dirS{"R", "U", "L", "D"};

int floodFillAlgo(int sr, int sc, int er, int ec, vector<vector<int>> &board, string ans)
{
    if (sr == er && sc == ec)
    {
        cout << ans << endl;
        return 1;
    }

    int count = 0;
    board[sr][sc] = 1; //mark visited
    for (int d = 0; d < dirA.size(); d++)
    {
        int r = sr + dirA[d][0];
        int c = sc + dirA[d][1];
        if (r >= 0 && c >= 0 && r <= er && c <= ec && board[r][c] == 0)
            count += floodFillAlgo(r, c, er, ec, board, ans + dirS[d]);
    }

    board[sr][sc] = 0; //mark unvisited
    return count;
}

int floodFillJumpAlgo(int sr, int sc, int er, int ec, vector<vector<int>> &board, string ans)
{
    if (sr == er && sc == ec)
    {
        cout << ans << endl;
        return 1;
    }

    int count = 0;
    board[sr][sc] = 1; //mark visited
    for (int d = 0; d < dirA.size(); d++)
    {
        for (int rad = 1; rad <= max(n, m); rad++)
        {
            int r = sr + rad * dirA[d][0];
            int c = sc + rad * dirA[d][1];
            if (r >= 0 && c >= 0 && r <= er && c <= ec && board[r][c] == 0)
                count += floodFillJumpAlgo(r, c, er, ec, board, ans + dirS[d] + to_string(rad));
        }
    }

    board[sr][sc] = 0; //mark unvisited
    return count;
}

class floodFillPair
{
public:
    string str = "";
    int len = 0;

    floodFillPair(string str, int len)
    {
        this->str = str;
        this->len = len;
    }
};

floodFillPair longestPath(int sr, int sc, int er, int ec, vector<vector<int>> &board)
{
    if (sr == er && sc == ec)
    {
        floodFillPair base("", 0);
        return base;
    }

    floodFillPair myAns("", 0);
    board[sr][sc] = 1; //mark visited
    for (int d = 0; d < dirA.size(); d++)
    {
        for (int rad = 1; rad <= max(n, m); rad++)
        {
            int r = sr + rad * dirA[d][0];
            int c = sc + rad * dirA[d][1];
            if (r >= 0 && c >= 0 && r <= er && c <= ec && board[r][c] == 0)
            {
                floodFillPair recAns = longestPath(r, c, er, ec, board);
                if (recAns.len + 1 > myAns.len)
                {
                    myAns.len = recAns.len + 1;
                    myAns.str = dirS[d] + to_string(rad) + recAns.str;
                }
            }
        }
    }

    board[sr][sc] = 0; //mark unvisited
    return myAns;
}

void solve()
{
    n = 3;
    m = 3;
    vector<vector<int>> board(n, vector<int>(m, 0));
    // cout << floodFillAlgo(0, 0, n - 1, m - 1, board, "") << endl;
    // cout << floodFillJumpAlgo(0, 0, n - 1, m - 1, board, "") << endl;
    cout << longestPath(0, 0, n - 1, m - 1, board).str << endl;
}

int main()
{
    solve();
    return 0;
}

File: 2019/LevelUpDecBatch/l001_Recursion/l002.java
import java.util.ArrayList;

public class l002 {

    public static void main(String[] args) {
        // System.out.println(printMazePaths(0, 0, 2, 2, ""));
        floodfill();
    }

    public static int printMazePaths(int sr, int sc, int dr, int dc, String psf) {
        if (sr == dr && sc == dc) {
            System.out.println(psf);
            return 1;
        }

        int count = 0;
        if (sc + 1 <= dc)
            count += printMazePaths(sr, sc + 1, dr, dc, psf + "h");
        if (sr + 1 <= dr && sc + 1 <= dc)
            count += printMazePaths(sr + 1, sc + 1, dr, dc, psf + "d");
        if (sr + 1 <= dr)
            count += printMazePaths(sr + 1, sc, dr, dc, psf + "v");

        return count;
    }

    public static int printMazePathsMultiJumps(int sr, int sc, int dr, int dc, String psf) {
        if (sr == dr && sc == dc) {
            System.out.println(psf);
            return 1;
        }

        int count = 0;
        for (int j = 1; sc + j <= dc; j++)
            count += printMazePathsMultiJumps(sr, sc + j, dr, dc, psf + "h" + j);
        for (int j = 1; sr + j <= dr && sc + j <= dc; j++)
            count += printMazePathsMultiJumps(sr + j, sc + j, dr, dc, psf + "d" + j);
        for (int j = 1; sr + j <= dr; j++)
            count += printMazePathsMultiJumps(sr + j, sc, dr, dc, psf + "v" + j);

        return count;
    }

    public static int floodFill(int sr, int sc, int er, int ec, boolean[][] vis, int[][] dir, String[] dirS,
            String ans) {
        if (sr == er && sc == ec) {
            System.out.println(ans);
            return 1;
        }

        int n = vis.length; // no of rows
        int m = vis[0].length; // no of col

        vis[sr][sc] = true; // mark

        int count = 0;
        for (int d = 0; d < dir.length; d++) { // call for all unvisited nbr's
            int r = sr + dir[d][0];
            int c = sc + dir[d][1];

            if (r >= 0 && c >= 0 && r < n && c < m && !vis[r][c])
                count += floodFill(r, c, er, ec, vis, dir, dirS, ans + dirS[d]);
        }

        vis[sr][sc] = false;

        return count;
    }

    public static int floodFillJump(int sr, int sc, int er, int ec, boolean[][] vis, int[][] dir, String[] dirS,
            String ans) {
        if (sr == er && sc == ec) {
            System.out.println(ans);
            return 1;
        }

        int n = vis.length; // no of rows
        int m = vis[0].length; // no of col

        vis[sr][sc] = true; // mark

        int count = 0;
        for (int d = 0; d < dir.length; d++) { // call for all unvisited nbr's
            for (int rad = 1; rad <= Math.max(n, m); rad++) {
                int r = sr + rad * dir[d][0];
                int c = sc + rad * dir[d][1];

                if (r >= 0 && c >= 0 && r < n && c < m) {
                    if (!vis[r][c])
                        count += floodFillJump(r, c, er, ec, vis, dir, dirS, ans + dirS[d] + rad);
                } else
                    break;
            }
        }

        vis[sr][sc] = false;

        return count;
    }

    public static class pair {
        int longestPathLength = 0;
        String longestPath = "";

        public pair(int longestPathLength, String longestPath) {
            this.longestPathLength = longestPathLength;
            this.longestPath = longestPath;
        }
    }

    public static pair longestPathRecu(int sr, int sc, int dr, int dc, boolean[][] vis, int[][] dir, String[] dirS) {
        if (sr == dr && sc == dc) {
            return new pair(0, "");
        }

        pair myAns = new pair(-1, "");
        vis[sr][sc] = true;

        for (int d = 0; d < dir.length; d++) {
            int r = sr + dir[d][0];
            int c = sc + dir[d][1];

            if (r >= 0 && c >= 0 && r <= dr && c <= dc && !vis[r][c]) {
                pair recAns = longestPathRecu(r, c, dr, dc, vis, dir, dirS);
                if (recAns.longestPathLength + 1 > myAns.longestPathLength) {
                    myAns.longestPathLength = recAns.longestPathLength + 1;
                    myAns.longestPath = dirS[d] + recAns.longestPath;
                }
            }
        }

        vis[sr][sc] = false;

        return myAns;
    }

    public static void floodfill() {
        boolean[][] vis = new boolean[3][3]; // vector<vector<bool>> vis(3,vector<bool>(3,false));
        int[][] dir = { { 1, 0 }, { -1, 0 }, { 1, 1 }, { -1, -1 }, { 0, 1 }, { 0, -1 }, { -1, 1 }, { 1, -1 } };
        String[] dirS = { "D", "U", "S", "N", "R", "L", "E", "W" };

        // int[][] dir = { { 1, 0 }, { 0, 1 }, { 1, 1 } };
        // String[] dirS = { "V", "H", "D" };
        System.out.println(floodFill(0, 0, 2, 2, vis, dir, dirS, ""));
    }

    // https://www.geeksforgeeks.org/rat-in-a-maze-backtracking-2/
    // https://www.geeksforgeeks.org/count-number-ways-reach-destination-maze/
    // https://www.geeksforgeeks.org/rat-in-a-maze-with-multiple-steps-jump-allowed/?ref=rp
    // https://www.geeksforgeeks.org/rat-in-a-maze-problem-when-movement-in-all-possible-directions-is-allowed/?ref=rp
}

    public static void fun(ArrayList<Integer> arr) {
        System.out.println(arr.get(0));
        int val = arr.get(0) + 10;
        arr.set(0, val);
        arr.set(0, val - 10);
    }

    public static void fun1() {
        ArrayList<Integer> arr = new ArrayList<>();
        arr.add(1);
        fun(arr);
        System.out.println(arr.get(0));

    }